### DISCUSSION TITLE
How to get the user id when registration form is used.

### DISCUSSION THREAD

**Discussion by JulieVelghe** [verbb/formie#1802] [Q&A]
I have a question about some custom development I'm doing.
We have a registration form that has an elements integration to create a user. The user has an entry field representing a chosen field of study. The data for that field is on the main site.
The form is on a subsite, and it has a field where the user can select an entry from the main site (being a chosen field of study). 

On the initial creating of the user, this field is not getting mapped and remains empty.

I have another form, to modify the user data. I used a custom module, where I checked in the event_after_submission, which field of study was chosen on that form and update the user data. That code works well:
```
                    if ($form->handle == "formUserData"){
                        $primarySiteId = Craft::$app->sites->primarySite->id;

                        //get the field userCourse and save it to the user's field userCourse
                        $submissionUserCourse = $submission->getFieldValue('userCourse')->siteId($primarySiteId)->one();
                        $submissionUserCourseId = $submissionUserCourse->id;
                        $submissionUserId = $submission->getFieldValue('userProfileId');

                        //find user
                        $user =  \craft\elements\User::find()->id($submissionUserId)->one();
                        $user->setFieldValue('userCourse', $submissionUserCourseId);
                        Craft::$app->elements->saveElement($user);
                    }
```
I was wondering if I could do something similar for the inititial registration form. I'm guessing that the user is only created after the submission event, so what event can I tap in to? Is there a way to get the id of the newly created user, AND keep my form data where the chosen field of study is kept?

**Comment by engram-design** [comment]
Just double-checking on my end, and a Formie Entries field mapped to a Craft Entries field is working for me. Just ensure that the enabled sources are the same (an entry picked in the form is allowed to be picked in the Craft entries form for the user).

But, like any integration, you can modify the payload that it sent to the provider before it's sent. In this case, it's not the traditional off-site API, it's Craft's element-saving function, but the concept still stands.

See [docs](https://verbb.io/craft-plugins/formie/docs/developers/events#the-beforeSendPayload-event).

```php
use verbb\formie\events\SendIntegrationPayloadEvent;
use verbb\formie\integrations\elements\User;
use yii\base\Event;

Event::on(User::class, User::EVENT_BEFORE_SEND_PAYLOAD, function(SendIntegrationPayloadEvent $event) {
    // $payload will be a UserElement, prepped and ready to be saved.
    $payload = $event->payload;
});
```

So while you can hook into this, you shouldn't _need_ to if you're mapping entries fields, and that should be working.

**Reply by JulieVelghe** [reply]
For some reason the mapping doesn't work for that field, So I've tried to hook into this event.
Unfortunately that doesn't seem to work either. Below is my code.

```
 Event::on(User::class, User::EVENT_BEFORE_SEND_PAYLOAD, function(SendIntegrationPayloadEvent $event) {
            // $payload will be a UserElement, prepped and ready to be saved.
            $payload = $event->payload;

            $submission = $event->submission;
            $form = \verbb\formie\elements\Form::find()
                ->id($submission->formId)
                ->one();

            if ($form) {
                if ($form->handle == "formUserRegistration") {
                    $primarySiteId = Craft::$app->sites->primarySite->id;

                    //get the field userCourse and save it to the user's field userCourse
                    $submissionUserCourse = $submission->getFieldValue('userCourse')->siteId($primarySiteId)->one();
                    $submissionUserCourseId = $submissionUserCourse->id;

                    $payload->setFieldValue('userCourse', $submissionUserCourse);

                    $event->payload = $payload;
                }
            }
        });
```
I get into the event, and I get into the last if-statement. The submission user course is being found, so there is an ID available. However, after I check the user, the userCourse field is still empty.

Am I setting this field in a wrong way?

**Reply by engram-design** [reply]
Quick note that you can get the form from a submission easier, and setting the value of an element field in Craft always needs to be an array of IDs. I've adjusted that below.

```php
Event::on(User::class, User::EVENT_BEFORE_SEND_PAYLOAD, function(SendIntegrationPayloadEvent $event) {
    // $payload will be a UserElement, prepped and ready to be saved.
    $payload = $event->payload;

    $submission = $event->submission;
    $form = $submission->getForm();

    if ($form && $form->handle == "formUserRegistration") {
        $primarySiteId = Craft::$app->sites->primarySite->id;

        //get the field userCourse and save it to the user's field userCourse
        $submissionUserCourse = $submission->getFieldValue('userCourse')->siteId($primarySiteId)->one();
        $submissionUserCourseId = $submissionUserCourse->id;

        $payload->setFieldValue('userCourse', [$submissionUserCourse]);

        $event->payload = $payload;
    }
});
